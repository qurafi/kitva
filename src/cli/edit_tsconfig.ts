import { readFile, writeFile } from "fs/promises";
import { warn } from "../utils/index.js";

const root_dirs = [".svelte-kit/types", ".schemas/types", "."];

function loadTsconfig(content: string) {
    if (content == "") return;
    try {
        return (0, eval)(`(${content})`);
    } catch (e) {
        return;
    }
}

const rootDir_key = '"rootDirs":';

export function editTsConfig(content: string) {
    const rootDirs_idx = content.indexOf(rootDir_key);
    if (rootDirs_idx >= 0) {
        return mergeRootDirs(content, rootDirs_idx);
    }

    const new_content = content.replace(/"compilerOptions": {([^"]*)/, (str, indent) => {
        return `${str}"rootDirs": ${JSON.stringify(root_dirs)},${indent}`;
    });

    if (!loadTsconfig(new_content)) {
        return content;
    }

    return new_content;
}

function mergeRootDirs(content: string, idx: number) {
    try {
        const rootDirs_json = content.slice(
            idx + rootDir_key.length,
            content.indexOf("]") + 1
        );
        const parsed = JSON.parse(rootDirs_json);
        const merged = [...new Set([...parsed, ...root_dirs])];
        return content.replace(rootDirs_json, JSON.stringify(merged));
    } catch (e) {
        return content;
    }
}

export async function addTsRootDir(config_path: string) {
    const content = await readFile(config_path, "utf-8");
    const tsconfig = loadTsconfig(content);
    if (!tsconfig) {
        throw new Error("invalid tsconfig syntax");
    }

    if (tsconfig?.compilerOptions?.rootDirs?.includes(".schemas/types")) {
        warn("Skipping adding rootDirs to tsconfig/jsconfig");
        return;
    }

    const new_content = editTsConfig(content);
    return writeFile(config_path, new_content);
}
