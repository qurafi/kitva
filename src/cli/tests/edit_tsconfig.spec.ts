import { describe, expect, it } from "vitest";
import { editTsConfig } from "../edit_tsconfig.js";

const sk_config = `{
	"compilerOptions": {
		"paths": {
			"$lib": [
				"../src/lib"
			],
			"$lib/*": [
				"../src/lib/*"
			]
		},
		"rootDirs": [
			"..",
			"./types"
		],
		"importsNotUsedAsValues": "error",
		"isolatedModules": true,
		"preserveValueImports": true,
		"lib": [
			"esnext",
			"DOM",
			"DOM.Iterable"
		],
		"moduleResolution": "node",
		"module": "esnext",
		"target": "esnext",
		"ignoreDeprecations": "5.0"
	},
	"include": [
		"ambient.d.ts",
		"./types/**/$types.d.ts",
		"../vite.config.ts",
		"../src/**/*.js",
		"../src/**/*.ts",
		"../src/**/*.svelte",
		"../tests/**/*.js",
		"../tests/**/*.ts",
		"../tests/**/*.svelte"
	],
	"exclude": [
		"../node_modules/**",
		"./[!ambient.d.ts]**",
		"../src/service-worker.js",
		"../src/service-worker.ts",
		"../src/service-worker.d.ts"
	]
}`;
describe("editTsconfig", () => {
	const fixtures = {
		empty: "",
		empty_obj: "{}",
		default_file: `{
        "extends": "./.svelte-kit/tsconfig.json",
        "compilerOptions": {
            "allowJs": true,
            "checkJs": true,
            "esModuleInterop": true,
            "forceConsistentCasingInFileNames": true,
            "resolveJsonModule": true,
            "skipLibCheck": true,
            "sourceMap": true,
            "strict": true
        }
        // Path aliases are handled by https://kit.svelte.dev/docs/configuration#alias
        //
        // If you want to overwrite includes/excludes, make sure to copy over the relevant includes/excludes
        // from the referenced tsconfig.json - TypeScript does not merge them in
    }
    `,

		with_root_dir_already: `{
        "extends": "./.svelte-kit/tsconfig.json",
        "compilerOptions": {
            "rootDirs": [".", ".svelte-kit/types", ".schemas/types"],
            "allowJs": true,
            "checkJs": true,
            "esModuleInterop": true,
            "forceConsistentCasingInFileNames": true,
            "resolveJsonModule": true,
            "skipLibCheck": true,
            "sourceMap": true,
            "strict": true
        }
        // Path aliases are handled by https://kit.svelte.dev/docs/configuration#alias
        //
        // If you want to overwrite includes/excludes, make sure to copy over the relevant includes/excludes
        // from the referenced tsconfig.json - TypeScript does not merge them in
    }
    `,

		with_root_dir_already_lines: `{
        "extends": "./.svelte-kit/tsconfig.json",
        "compilerOptions": {
            "rootDirs": [
                ".",
                ".svelte-kit/types",
                ".schemas/types",
                "userRootDir",
            ],
            "allowJs": true,
            "checkJs": true,
            "esModuleInterop": true,
            "forceConsistentCasingInFileNames": true,
            "resolveJsonModule": true,
            "skipLibCheck": true,
            "sourceMap": true,
            "strict": true
        }
        // Path aliases are handled by https://kit.svelte.dev/docs/configuration#alias
        //
        // If you want to overwrite includes/excludes, make sure to copy over the relevant includes/excludes
        // from the referenced tsconfig.json - TypeScript does not merge them in
    }
    `,
		invalid_syntax: `{
        "extends": "./.svelte-kit/tsconfig.json",
        "compilerOptions": {
            "rootDirs": [
                ".",
                ".svelte-kit/types",
                ".schemas/types",
                "userRootDir",
            ,
            "allowJs": true,
            "checkJs": true,
            "esModuleInterop": true,
            "forceConsistentCasingInFileNames": true,
            "resolveJsonModule": true,
            "skipLibCheck": true,
            "sourceMap": true,
            "strict": true
        }
        // Path aliases are handled by https://kit.svelte.dev/docs/configuration#alias
        //
        // If you want to overwrite includes/excludes, make sure to copy over the relevant includes/excludes
        // from the referenced tsconfig.json - TypeScript does not merge them in
    }
    `,

		with_include_another_thing: `{
    "extends": "./.svelte-kit/tsconfig.json",
    "compilerOptions": {
        "rootDirs": [".", ".svelte-kit/types", ".schemas/types"],
        "allowJs": true,
        "checkJs": true,
        "esModuleInterop": true,
        "forceConsistentCasingInFileNames": true,
        "resolveJsonModule": true,
        "skipLibCheck": true,
        "sourceMap": true,
        "strict": true
    },
    "include": ["something"]
    // Path aliases are handled by https://kit.svelte.dev/docs/configuration#alias
    //
    // If you want to overwrite includes/excludes, make sure to copy over the relevant includes/excludes
    // from the referenced tsconfig.json - TypeScript does not merge them in
}
`,
		with_include_another_thing_no_root_dirs: `{
    "extends": "./.svelte-kit/tsconfig.json",
    "compilerOptions": {
        "allowJs": true,
        "checkJs": true,
        "esModuleInterop": true,
        "forceConsistentCasingInFileNames": true,
        "resolveJsonModule": true,
        "skipLibCheck": true,
        "sourceMap": true,
        "strict": true
    },
    "include": ["something"]
    // Path aliases are handled by https://kit.svelte.dev/docs/configuration#alias
    //
    // If you want to overwrite includes/excludes, make sure to copy over the relevant includes/excludes
    // from the referenced tsconfig.json - TypeScript does not merge them in
}
`,

		already_copied: `{
    "extends": "./.svelte-kit/tsconfig.json",
    "compilerOptions": {
        "rootDirs": [
            ".",
            ".svelte-kit/types",
            ".schemas/types",
        ],
        "allowJs": true,
        "checkJs": true,
        "esModuleInterop": true,
        "forceConsistentCasingInFileNames": true,
        "resolveJsonModule": true,
        "skipLibCheck": true,
        "sourceMap": true,
        "strict": true
    },
    "include": [
        ".svelte-kit/ambient.d.ts",
        ".svelte-kit/types/**/$types.d.ts",
        "vite.config.ts",
        "src/**/*.js",
        "src/**/*.ts",
        "src/**/*.svelte",
        "src/**/*.js",
        "src/**/*.ts",
        "src/**/*.svelte",
        "tests/**/*.js",
        "tests/**/*.ts",
        "tests/**/*.svelte"
    ]
    // Path aliases are handled by https://kit.svelte.dev/docs/configuration#alias
    //
    // If you want to overwrite includes/excludes, make sure to copy over the relevant includes/excludes
    // from the referenced tsconfig.json - TypeScript does not merge them in
}
`,
		already_copied_but_modified: `{
    "extends": "./.svelte-kit/tsconfig.json",
    "compilerOptions": {
        "rootDirs": [
            ".",
            ".svelte-kit/types",
            ".schemas/types",
            "userRootDir"
        ],
        "allowJs": true,
        "checkJs": true,
        "esModuleInterop": true,
        "forceConsistentCasingInFileNames": true,
        "resolveJsonModule": true,
        "skipLibCheck": true,
        "sourceMap": true,
        "strict": true
    },
    "include": [
        ".svelte-kit/ambient.d.ts",
        ".svelte-kit/types/**/$types.d.ts",
        "vite.config.ts",
        "src/**/*.js",
        "src/**/*.ts",
        "src/**/*.svelte",
        "src/**/*.js",
        "src/**/*.ts",
        "src/**/*.svelte",
        "tests/**/*.js",
        "tests/**/*.ts",
        "tests/**/*.svelte"
    ]
    // Path aliases are handled by https://kit.svelte.dev/docs/configuration#alias
    //
    // If you want to overwrite includes/excludes, make sure to copy over the relevant includes/excludes
    // from the referenced tsconfig.json - TypeScript does not merge them in
}
`
	};

	for (const [name, content] of Object.entries(fixtures)) {
		it(`should edit tsconfig config correctly: ${name}`, async () => {
			const result = await editTsConfig(content, sk_config);
			// console.log(name, content, "\n\n", result);
			expect(result).toMatchSnapshot();
		});
	}
});
