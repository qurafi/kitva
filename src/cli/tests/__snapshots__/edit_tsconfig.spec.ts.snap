// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`editTsconfig > should edit tsconfig config correctly: already_copied 1`] = `
"{
    \\"extends\\": \\"./.svelte-kit/tsconfig.json\\",
    \\"compilerOptions\\": {
        \\"rootDirs\\": [
            \\".\\",
            \\".svelte-kit/types\\",
            \\".schemas/types\\"
        ],
        \\"allowJs\\": true,
        \\"checkJs\\": true,
        \\"esModuleInterop\\": true,
        \\"forceConsistentCasingInFileNames\\": true,
        \\"resolveJsonModule\\": true,
        \\"skipLibCheck\\": true,
        \\"sourceMap\\": true,
        \\"strict\\": true
    },
    \\"include\\": [
        \\".svelte-kit/ambient.d.ts\\",
        \\".svelte-kit/types/**/$types.d.ts\\",
        \\"vite.config.ts\\",
        \\"src/**/*.js\\",
        \\"src/**/*.ts\\",
        \\"src/**/*.svelte\\",
        \\"src/**/*.js\\",
        \\"src/**/*.ts\\",
        \\"src/**/*.svelte\\",
        \\"tests/**/*.js\\",
        \\"tests/**/*.ts\\",
        \\"tests/**/*.svelte\\",
        \\".schemas/types\\"
    ]
    // Path aliases are handled by https://kit.svelte.dev/docs/configuration#alias
    //
    // If you want to overwrite includes/excludes, make sure to copy over the relevant includes/excludes
    // from the referenced tsconfig.json - TypeScript does not merge them in
}
"
`;

exports[`editTsconfig > should edit tsconfig config correctly: already_copied_but_modified 1`] = `
"{
    \\"extends\\": \\"./.svelte-kit/tsconfig.json\\",
    \\"compilerOptions\\": {
        \\"rootDirs\\": [
            \\".\\",
            \\".svelte-kit/types\\",
            \\".schemas/types\\",
            \\"userRootDir\\"
        ],
        \\"allowJs\\": true,
        \\"checkJs\\": true,
        \\"esModuleInterop\\": true,
        \\"forceConsistentCasingInFileNames\\": true,
        \\"resolveJsonModule\\": true,
        \\"skipLibCheck\\": true,
        \\"sourceMap\\": true,
        \\"strict\\": true
    },
    \\"include\\": [
        \\".svelte-kit/ambient.d.ts\\",
        \\".svelte-kit/types/**/$types.d.ts\\",
        \\"vite.config.ts\\",
        \\"src/**/*.js\\",
        \\"src/**/*.ts\\",
        \\"src/**/*.svelte\\",
        \\"src/**/*.js\\",
        \\"src/**/*.ts\\",
        \\"src/**/*.svelte\\",
        \\"tests/**/*.js\\",
        \\"tests/**/*.ts\\",
        \\"tests/**/*.svelte\\",
        \\".schemas/types\\"
    ]
    // Path aliases are handled by https://kit.svelte.dev/docs/configuration#alias
    //
    // If you want to overwrite includes/excludes, make sure to copy over the relevant includes/excludes
    // from the referenced tsconfig.json - TypeScript does not merge them in
}
"
`;

exports[`editTsconfig > should edit tsconfig config correctly: default_file 1`] = `
"{
        \\"extends\\": \\"./.svelte-kit/tsconfig.json\\",
        \\"compilerOptions\\": {
            \\"allowJs\\": true,
            \\"checkJs\\": true,
            \\"esModuleInterop\\": true,
            \\"forceConsistentCasingInFileNames\\": true,
            \\"resolveJsonModule\\": true,
            \\"skipLibCheck\\": true,
            \\"sourceMap\\": true,
            \\"strict\\": true,
            \\"rootDirs\\": [
                \\".\\",
                \\".svelte-kit/types\\",
                \\".schemas/types\\"
            ]
        },
        \\"include\\": [
            \\".svelte-kit/ambient.d.ts\\",
            \\".svelte-kit/types/**/$types.d.ts\\",
            \\"vite.config.ts\\",
            \\"src/**/*.js\\",
            \\"src/**/*.ts\\",
            \\"src/**/*.svelte\\",
            \\"tests/**/*.js\\",
            \\"tests/**/*.ts\\",
            \\"tests/**/*.svelte\\",
            \\".schemas/types\\"
        ]
        // Path aliases are handled by https://kit.svelte.dev/docs/configuration#alias
        //
        // If you want to overwrite includes/excludes, make sure to copy over the relevant includes/excludes
        // from the referenced tsconfig.json - TypeScript does not merge them in
    }
    "
`;

exports[`editTsconfig > should edit tsconfig config correctly: empty 1`] = `undefined`;

exports[`editTsconfig > should edit tsconfig config correctly: empty_obj 1`] = `
"{
    \\"include\\": [
        \\".svelte-kit/ambient.d.ts\\",
        \\".svelte-kit/types/**/$types.d.ts\\",
        \\"vite.config.ts\\",
        \\"src/**/*.js\\",
        \\"src/**/*.ts\\",
        \\"src/**/*.svelte\\",
        \\"tests/**/*.js\\",
        \\"tests/**/*.ts\\",
        \\"tests/**/*.svelte\\",
        \\".schemas/types\\"
    ],
    \\"compilerOptions\\": {
        \\"rootDirs\\": [
            \\".\\",
            \\".svelte-kit/types\\",
            \\".schemas/types\\"
        ]
    }
}"
`;

exports[`editTsconfig > should edit tsconfig config correctly: invalid_syntax 1`] = `undefined`;

exports[`editTsconfig > should edit tsconfig config correctly: with_include_another_thing 1`] = `
"{
    \\"extends\\": \\"./.svelte-kit/tsconfig.json\\",
    \\"compilerOptions\\": {
        \\"rootDirs\\": [
            \\".\\",
            \\".svelte-kit/types\\",
            \\".schemas/types\\"
        ],
        \\"allowJs\\": true,
        \\"checkJs\\": true,
        \\"esModuleInterop\\": true,
        \\"forceConsistentCasingInFileNames\\": true,
        \\"resolveJsonModule\\": true,
        \\"skipLibCheck\\": true,
        \\"sourceMap\\": true,
        \\"strict\\": true
    },
    \\"include\\": [
        \\"something\\",
        \\".schemas/types\\"
    ]
    // Path aliases are handled by https://kit.svelte.dev/docs/configuration#alias
    //
    // If you want to overwrite includes/excludes, make sure to copy over the relevant includes/excludes
    // from the referenced tsconfig.json - TypeScript does not merge them in
}
"
`;

exports[`editTsconfig > should edit tsconfig config correctly: with_include_another_thing_no_root_dirs 1`] = `
"{
    \\"extends\\": \\"./.svelte-kit/tsconfig.json\\",
    \\"compilerOptions\\": {
        \\"allowJs\\": true,
        \\"checkJs\\": true,
        \\"esModuleInterop\\": true,
        \\"forceConsistentCasingInFileNames\\": true,
        \\"resolveJsonModule\\": true,
        \\"skipLibCheck\\": true,
        \\"sourceMap\\": true,
        \\"strict\\": true,
        \\"rootDirs\\": [
            \\".\\",
            \\".svelte-kit/types\\",
            \\".schemas/types\\"
        ]
    },
    \\"include\\": [
        \\"something\\",
        \\".schemas/types\\"
    ]
    // Path aliases are handled by https://kit.svelte.dev/docs/configuration#alias
    //
    // If you want to overwrite includes/excludes, make sure to copy over the relevant includes/excludes
    // from the referenced tsconfig.json - TypeScript does not merge them in
}
"
`;

exports[`editTsconfig > should edit tsconfig config correctly: with_root_dir_already 1`] = `
"{
        \\"extends\\": \\"./.svelte-kit/tsconfig.json\\",
        \\"compilerOptions\\": {
            \\"rootDirs\\": [
                \\".\\",
                \\".svelte-kit/types\\",
                \\".schemas/types\\"
            ],
            \\"allowJs\\": true,
            \\"checkJs\\": true,
            \\"esModuleInterop\\": true,
            \\"forceConsistentCasingInFileNames\\": true,
            \\"resolveJsonModule\\": true,
            \\"skipLibCheck\\": true,
            \\"sourceMap\\": true,
            \\"strict\\": true
        },
        \\"include\\": [
            \\".svelte-kit/ambient.d.ts\\",
            \\".svelte-kit/types/**/$types.d.ts\\",
            \\"vite.config.ts\\",
            \\"src/**/*.js\\",
            \\"src/**/*.ts\\",
            \\"src/**/*.svelte\\",
            \\"tests/**/*.js\\",
            \\"tests/**/*.ts\\",
            \\"tests/**/*.svelte\\",
            \\".schemas/types\\"
        ]
        // Path aliases are handled by https://kit.svelte.dev/docs/configuration#alias
        //
        // If you want to overwrite includes/excludes, make sure to copy over the relevant includes/excludes
        // from the referenced tsconfig.json - TypeScript does not merge them in
    }
    "
`;

exports[`editTsconfig > should edit tsconfig config correctly: with_root_dir_already_lines 1`] = `
"{
        \\"extends\\": \\"./.svelte-kit/tsconfig.json\\",
        \\"compilerOptions\\": {
            \\"rootDirs\\": [
                \\".\\",
                \\".svelte-kit/types\\",
                \\".schemas/types\\",
                \\"userRootDir\\"
            ],
            \\"allowJs\\": true,
            \\"checkJs\\": true,
            \\"esModuleInterop\\": true,
            \\"forceConsistentCasingInFileNames\\": true,
            \\"resolveJsonModule\\": true,
            \\"skipLibCheck\\": true,
            \\"sourceMap\\": true,
            \\"strict\\": true
        },
        \\"include\\": [
            \\".svelte-kit/ambient.d.ts\\",
            \\".svelte-kit/types/**/$types.d.ts\\",
            \\"vite.config.ts\\",
            \\"src/**/*.js\\",
            \\"src/**/*.ts\\",
            \\"src/**/*.svelte\\",
            \\"tests/**/*.js\\",
            \\"tests/**/*.ts\\",
            \\"tests/**/*.svelte\\",
            \\".schemas/types\\"
        ]
        // Path aliases are handled by https://kit.svelte.dev/docs/configuration#alias
        //
        // If you want to overwrite includes/excludes, make sure to copy over the relevant includes/excludes
        // from the referenced tsconfig.json - TypeScript does not merge them in
    }
    "
`;
